main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala

```diff
@@ -360,7 +360,12 @@ object ConstraintSolver {
         res <- CaseSetUnification.unify(t1, t2, renv, sym1.universe, sym1).mapErr(toTypeError(_, prov))
       } yield ResolutionResult.newSubst(res)
 
-    case (k1, k2) if KindUnification.unifiesWith(k1, k2) => resolveEqualityStar(tpe1, tpe2, prov, renv, loc)
+    case (k1, k2) if KindUnification.unifiesWith(k1, k2) =>
+      for {
+        (t1, _) <- TypeReduction.simplify(tpe1, renv, loc)
+        (t2, _) <- TypeReduction.simplify(tpe2, renv, loc)
+        res <- resolveEqualityStar(tpe1, tpe2, prov, renv, loc)
```

Comment:
We need to turn understand the performance implications of this.

Can you run `Xperf --frontend n 21` on master and on your code, and report the results back here.

Jonathan can help with the commands, if there are some issues.